### DISCUSSION TITLE
Is it safe to call JNI in the recordSample function?

### DISCUSSION THREAD

**Discussion by hustchaya** [async-profiler/async-profiler#811] [Q&A] [answered]
I want to analyze the classloader of the sampled thread, I try not to call JVMTI directly, I have got "thread_info.context_class_loader" . Further parsing needs to use JNI calls, I have tried direct or indirect calls, but all failed. Is there any method you can recommend?

**Comment by hustchaya** [comment]
call JVMTI  not directly by "_libjvm->findSymbolByPrefix("_ZN8JvmtiEnv16GetCurrentThreadEPP8_jobject")"
call JVMTI  not directly by "_libjvm->findSymbolByPrefix("_ZN8JvmtiEnv13GetThreadInfoEP8_jobjectP16_jvmtiThreadInfo")"

**Comment by apangin** [comment, marked as answer]
Neither JNI nor JVM TI is safe inside a signal handler.
JNI and JVM TI can lock, allocate, and call all sorts of other non-async-signal-safe functions.
In general, it's unsafe to access anything on Java heap inside `recordSample` - imagine it can run concurrently with GC, for example.

**Reply by hustchaya** [reply]
![image](https://github.com/async-profiler/async-profiler/assets/102569465/a3123c88-4a2c-423f-b2cd-dbfaac38f533)
I see the annotation as shown above，It is safe to call JVMTI indirectly.Is it safe to call JVMTI indirectly under the condition of event_type == 0??

**Reply by apangin** [reply]
I think I answered the question above. In short - no, it's unsafe.

**Comment by hustchaya** [comment]
My main goal is to distinguish the CPU and MEM usage of different claasloaders(thread-based). I want to analyze the contextclassloader of the thread when sampling. Is this feasible? Or can I trouble you to give me a better suggestion?

**Reply by apangin** [reply]
I'm not sure what you want to achieve by matching profile data with class loaders (especially with context class loaders which are kind of exotic thing), but in any case, you may record context class loaders separately: on thread start and on `setContextClassLoader` call, and later match them with profiling events by a timestamp.

**Reply by hustchaya** [reply]
I understand, thanks for the reply
